added appendix and problem section (#18)
MrLongbottom/P9-Rapport: sections/appendix/problems.tex

\subsection{Encountered Problems}

\subsubsection{Too many low values}
The document-topic distributions seem to have too low entropy and a lot of very low values.

> This causes each topic to have an average of $\sim$7000 docs in their distribution, however with very low entropy, meaning that some of these distributions are very insignificant, but add space and calculation time to our sparse representations.

Reviewer: ```suggestion
This causes each topic to have an average of $\sim$7000 docs in their distribution, however with very low entropy, meaning that some of these distributions are very insignificant, but add space and time complexity to our sparse representations.
```

Author: ```suggestion
This causes each topic to have an average of $\sim$7000 docs in their distribution, however with very low entropy, meaning that some of these distributions are very insignificant, but still add space and time complexity to our sparse representations.
```

To fix this we try tuning a minimum threshold for a topic to be counted in the document-topic distributions, which can be set as a hyper-parameter for \emph{gensim.ldamodel.get\_document\_topics()}.

> After setting up a minimum percent of 2.5\% for a entry to be included in the document distribution, the amount of documents per topic fell to ~4800, removing more than 5 entries per document for a total of $\sim$350.000 entries deleted. This has overall made the matrices a lot more sparse, but we will still have to evaluate whether changing K is better.

Reviewer: ```suggestion
After setting up a minimum percent of 2.5\% for an entry to be included in the document distribution, the number of documents per topic fell to ~4800, removing more than 5 entries per document for a total of $\sim$350.000 entries deleted. 
This has overall made the matrices a lot more sparse, but we will still have to evaluate whether changing K is better.
```

\subsubsection{Empty Topics}

> There are also a significant number of topics, which had words, but no assigned docs. These should be detected and cut away, though their existence points towards a lower $K$ might be needed.

Reviewer: ```suggestion
There are also a significant number of topics, which had words, but no assigned docs. 
These should be detected and removed, though their existence points towards a lower $K$ might be needed.
```

\subsubsection{Too Common Topics}

> On the opposite end we also have some topics which are so universal that almost all documents are included in this topic. This might be a sign that pre-processing is not working well enough, or it might once again point at the needed for a better K. Alternatively, too general topics could be manually pruned after the LDA model has been run.

Reviewer: ```suggestion
On the opposite end, we also have some topics which are so universal that almost all documents are included in this topic. 
This might be a sign that pre-processing is not working well enough, or it might once again point at the needed for a better K.
Alternatively, too general topics could be manually pruned after the LDA model has been run.
```

\subsubsection{Adjacency Matrix Construction}

> Our adjacency matrix is based on whether documents share some topics in their topic distributions. This lead us to the problem of D\*D\*T insertion operations into a sparse matrix with what was originally ~50.000 documents. (unknown)

Reviewer: ```suggestion
Our adjacency matrix is based on whether documents share some topics in their topic distributions. 
This leads us to the problem of D\*D\*T insertion operations into a sparse matrix with what was originally ~50.000 documents. (unknown)
```

We did a lot of prepossessing, which reduced this number eventually all the way down to ~30.000 documents, but it was too insignificant to solve anything. (~140 days)
However our similarity function is symmetric, so we cut the time in half by only constructing half of the adjacency matrix. (~69 days)
We then changed our algorithm to only consider documents that shared topics to the original documents, decreasing our time to D\*T\_d\*D\_t, where both of the new variables are reduced compared to the originals. (~48 days)
We also made the calculation multithreaded in order to reduce compution time (~2,4 days)

> Lastly we changed to construct our matrix using the \emph{lil\_matrix} format, which has faster insertions.

Reviewer: ```suggestion
Lastly, we changed to construct our matrix using the \emph{lil\_matrix} format, which has faster insertions.
```

> The function get\_document\_topics might be needing the whole corpus since whenever we give it the query it returns the same two topics. These two topics are probably the two most general topics within the model which might spell trouble since we don't know how to remove these from the model.

Reviewer: ```suggestion
The function get\_document\_topics might be needing the whole corpus since whenever we give it the query it returns the same two topics. 
These two topics are probably the two most general topics within the model which might spell trouble since we don't know how to remove these from the model.
```

\subsubsection{Dirichlets produce Non-Zero values}

> We use distributions produced by dirichlets to construct our adjacency matrix and to convert queries into topic distributions. However as part of how lda is designed it never has 0 values in its distribution values, only extremely low numbers. This is not optimal for constructing an adjacency matrix, as it would result in a fully connected graph.

Reviewer: ```suggestion
We use distributions produced by the Dirichlet's to construct our adjacency matrix and to convert queries into topic distributions. 
However as part of how LDA is designed it never has 0 values in its distribution values, only extremely low numbers.
This is not optimal for constructing an adjacency matrix, as it would result in a fully connected graph.
```

Adjusting $\alpha$ and $\eta$ will make the low values lower, and the high values higher, but the problem never disappears from this.

> We have therefore introduced thresholds. These thresholds reduce all values lower than the threshold to zero in the document-topic matrix and topic-word matrix.

Reviewer: ```suggestion
We have therefore introduced thresholds. 
These thresholds reduce all values lower than the threshold to zero in the document-topic matrix and topic-word matrix.
```

> The new problem introduced by introducing thresholds, is that the exact values of the thresholds are now added as extra hyper-parameters for our solution.

Reviewer: ```suggestion
The new problem introduced by introducing thresholds is that the exact values of the thresholds are now added as extra hyper-parameters for our solution.
```

> \subsection{Encountered Problems}

Reviewer 2: Vi skal nok være tydeligere i om det er problemer vi fik løst, eller bare har opdaget, da det ikke altid er tydeligt. Kan måske først gøres når vi nærmer os en færdig implementation.

Author: ```suggestion
\subsection{Encountered Problems}
\todo{indicate whether the problems were solved.}
```

\subsubsection{Too many low values}
The document-topic distributions seem to have too low entropy and a lot of very low values.
This causes each topic to have an average of $\sim$7000 docs in their distribution, however with very low entropy, meaning that some of these distributions are very insignificant, but still add space and time complexity to our sparse representations.
To fix this we try tuning a minimum threshold for a topic to be counted in the document-topic distributions, which can be set as a hyper-parameter for \emph{gensim.ldamodel.get\_document\_topics()}.

> After setting up a minimum percent of 2.5\% for an entry to be included in the document distribution, the number of documents per topic fell to ~4800, removing more than 5 entries per document for a total of $\sim$350.000 entries deleted. 

Author: Hele dette problem kommer tilbage til at vi ikke havet forstået at de distributions vi ville få ud kom som standard ikke til at have 0'ere i sig. Ville mene at dette problem skal omskrives, fjernes, eller kombineres med at snakke om at lave thresholds til topic-word distribution også.
Tror måske jeg bare fjerner den fra dette PR, da jeg også bare gerne vil have nogle af disse problem beskrivelser sendt videre til vejledere (de lyder til gerne at vil se dem)

This has overall made the matrices a lot more sparse, but we will still have to evaluate whether changing K is better.

\subsubsection{Empty Topics}

> There are also a significant number of topics, which had words, but no assigned docs. 

Author: dette problem eksisterer vidst ikke længere, så vidt jeg husker. Kan ikke huske hvordan vi fjernede det dog.

> These should be detected and removed, though their existence points towards a lower $K$ might be needed.

Reviewer 2: ```suggestion
These should be detected and removed, though their existence points towards a lower K might be needed.
```

Author: kommer K ikke til at se anderledes ud når det ikke er i math mode?

Reviewer 2: Jo, det var bare det eneste K i math mode. Men afsnittet er alligevel væk indtil videre.

\subsubsection{Too Common Topics}
On the opposite end, we also have some topics which are so universal that almost all documents are included in this topic. 

> This might be a sign that pre-processing is not working well enough, or it might once again point at the needed for a better K.

Reviewer 2: ```suggestion
This might be a sign that the preprocessing is not working well enough, or it might once again point at the need for a better K.
```

Alternatively, too general topics could be manually pruned after the LDA model has been run.

\subsubsection{Adjacency Matrix Construction}
Our adjacency matrix is based on whether documents share some topics in their topic distributions. 

> This leads us to the problem of D\*D\*T insertion operations into a sparse matrix with what was originally ~50.000 documents. (unknown)

Reviewer 2: ```suggestion
This leads us to the problem of $D*D*T$ insertion operations into a sparse matrix with what was originally ~50.000 documents. (unknown)
```

> This leads us to the problem of D\*D\*T insertion operations into a sparse matrix with what was originally ~50.000 documents. (unknown)

Reviewer 2: Jeg havde umiddelbart ingen anelse hvad (unknown) betød, men det er vidst tiden insertion ville tage. Det skal beskrives.

> We did a lot of prepossessing, which reduced this number eventually all the way down to ~30.000 documents, but it was too insignificant to solve anything. (~140 days)

Reviewer 2: ```suggestion
We did a lot of preprocessing, which reduced this number eventually all the way down to ~30.000 documents, but it was too insignificant to solve anything. (~140 days)
```

However our similarity function is symmetric, so we cut the time in half by only constructing half of the adjacency matrix. (~69 days)
We then changed our algorithm to only consider documents that shared topics to the original documents, decreasing our time to D\*T\_d\*D\_t, where both of the new variables are reduced compared to the originals. (~48 days)

> We also made the calculation multithreaded in order to reduce compution time (~2,4 days)

Reviewer 2: ```suggestion
We also made the calculation multithreaded in order to reduce computation time. (~2,4 days)
```

> Lastly, we changed to construct our matrix using the \emph{lil\_matrix} format, which has faster insertions.

Reviewer 2: Hvad er tiden efter det?

> The function get\_document\_topics might be needing the whole corpus since whenever we give it the query it returns the same two topics. 

Reviewer 2: Hvorfor nævner vi `get\_document\_topics`? Er det en Gensim funktion eller vores egen? Det er ikke tydeligt hvad det har med adjacency matrices at gøre.

Author: tror måske dette er meningen det skal være et seperat problem. Skulle ikke mene det er mig som har skrevet dette.

Reviewer 2: ```suggestion
\todo[inline]{Nedenstående problem skal uddybes, hvis det beholdes. Skal også have sit eget afsnit separat fra Adjacency Matrix Construction}
The function get\_document\_topics might be needing the whole corpus since whenever we give it the query it returns the same two topics. 
```

These two topics are probably the two most general topics within the model which might spell trouble since we don't know how to remove these from the model.

\subsubsection{Dirichlets produce Non-Zero values}

> We use distributions produced by the Dirichlet's to construct our adjacency matrix and to convert queries into topic distributions. 

Reviewer 2: ```suggestion
We use distributions produced by the Dirichlets to construct our adjacency matrix and to convert queries into topic distributions. 
```

> However as part of how LDA is designed it never has 0 values in its distribution values, only extremely low numbers.

Reviewer 2: ```suggestion
However, as part of how \gls{lda} is designed it never has 0 values in its topics word distribution, only extremely low numbers.
```

Author: det er begge distributions ikke bare topic-word

This is not optimal for constructing an adjacency matrix, as it would result in a fully connected graph.
Adjusting $\alpha$ and $\eta$ will make the low values lower, and the high values higher, but the problem never disappears from this.
We have therefore introduced thresholds. 

> These thresholds reduce all values lower than the threshold to zero in the document-topic matrix and topic-word matrix.

Reviewer 2: En ting der måske ikke skal løses her i appendix, men nok bare kan laves som et Github issue hvis det ikke er gjort: Beskriver vi noget sted at document-topic matrix er det samme som theta, og at topic-word matrix er det samme som beta?

Author: sure, lav et issue.

\subsection{Encountered Problems}
\todo[inline]{indicate whether the problems were solved.}

\subsubsection{Adjacency Matrix Construction}
Our adjacency matrix is based on whether documents share some topics in their topic distributions. 

> This leads us to the problem of $D*D*T$ insertion operations into a sparse matrix with what was originally ~50.000 documents. We never tested how long this would end up taking, but it was definitly infeasable.

Reviewer 2: ```suggestion
This leads us to the problem of $D*D*T$ insertion operations into a sparse matrix with what was originally ~50.000 documents. We never tested how long this would end up taking, but it was definitely infeasible.
```

We did a lot of preprocessing, which reduced this number eventually all the way down to ~30.000 documents, but it was too insignificant to solve anything. Calculation time was now estimated to be ~140 days.
However our similarity function is symmetric, so we cut the time in half by only constructing half of the adjacency matrix. Calculation time was now estimated to be ~69 days.

> We then changed our algorithm to only consider documents that shared topics to the original documents, decreasing our time to D\*T\_d\*D\_t, where both of the new variables are reduced compared to the originals. This reduced to calculation time to be estimated as ~48 days.

Reviewer 2: ```suggestion
We then changed our algorithm to only consider documents that shared topics to the original documents, decreasing our time to $D*T_d*D_t$, where both of the new variables are reduced compared to the originals. This reduced the calculation time to be estimated as ~48 days.
```

We also made the calculation multithreaded in order to reduce computation time. Estimated calculation time is reduced to ~2,4 days using 8 threads.
Lastly, we changed to construct our matrix using the \emph{lil\_matrix} format, which has faster insertions. This last change made the construction much more manageable.

The function get\_document\_topics might be needing the whole corpus since whenever we give it the query it returns the same two topics. 
These two topics are probably the two most general topics within the model which might spell trouble since we don't know how to remove these from the model.

> \subsubsection{Dirichlets only produce non-zero values}
> We use document-topic and topic-word distributions produced by the Dirichlets to construct our adjacency matrix and to convert queries into topic distributions. 

Reviewer 2: ```suggestion
\subsubsection{Dirichlet distributions only produce non-zero values}
We use document-topic and topic-word distributions produced by the Dirichlet distributions to construct our adjacency matrix and to convert queries into topic distributions. 
```
Jeg har fået set LDA videoerne, og umiddelbart kan man ikke kalde dem "Dirichlets".

However as part of how \gls{lda} is designed it has no zeros in its $\theta$ and $\beta$ values, only extremely low numbers.
This is not optimal for constructing an adjacency matrix, as it would result in a fully connected graph.

> Adjusting $\alpha$ and $\eta$ will make the low values lower, and the high values higher, but this does not make the problem disappears.

Reviewer 2: ```suggestion
Adjusting $\alpha$ and $\eta$ will make the low values lower, and the high values higher, but this does not make the problem disappear.
```
